pkg/connector/service-principals.go

```diff
@@ -207,6 +200,177 @@ func (s *servicePrincipalBuilder) Grants(ctx context.Context, resource *v2.Resou
+				l.Info(
+					"databricks-connector: service principal already has the entitlement",
+					zap.String("principal_id", principalID),
+					zap.String("entitlement", entitlement.Slug),
+				)
+
+				return nil, nil
+			}
+
+			// add the principal to the rule set
+			ruleSets[i].Principals = append(ruleSets[i].Principals, principalID)
+		}
+	}
+
+	if len(ruleSets) == 0 || !found {
+		ruleSets = append(ruleSets, databricks.RuleSet{
+			Role:       entitlement.Slug,
+			Principals: []string{principalID},
+		})
+	}
+
+	err = s.client.UpdateRuleSets(ctx, ServicePrincipalsType, applicationId, ruleSets)
+	if err != nil {
+		return nil, fmt.Errorf("databricks-connector: failed to update rule sets for service principal %s (%s): %w", principal.Id.Resource, applicationId, err)
+	}
+
+	return nil, nil
+}
+
+func (s *servicePrincipalBuilder) Revoke(ctx context.Context, grant *v2.Grant) (annotations.Annotations, error) {
+	l := ctxzap.Extract(ctx)
+
+	principal := grant.Principal
+	entitlement := grant.Entitlement
+
+	if !isValidPrincipal(principal.Id) {
+		l.Warn(
+			"databricks-connector: only users, groups and service principals can have service principal permissions revoked",
+			zap.String("principal_id", principal.Id.String()),
+			zap.String("principal_type", principal.Id.ResourceType),
+		)
+
+		return nil, fmt.Errorf("databricks-connector: only users, groups and service principals can have service principal permissions revoked")
+	}
+
+	groupTrait, err := rs.GetGroupTrait(entitlement.Resource)
+	if err != nil {
+		return nil, fmt.Errorf("databricks-connector: failed to get group trait: %w", err)
+	}
+
+	parentType, parentID, err := getParentInfoFromProfile(groupTrait.Profile)
+	if err != nil {
+		return nil, fmt.Errorf("databricks-connector: failed to get parent info from group profile: %w", err)
+	}
+
+	if parentType == workspaceResourceType.Id {
+		s.client.SetWorkspaceConfig(parentID)
+	} else {
+		s.client.SetAccountConfig()
+	}
+
+	applicationId, ok := rs.GetProfileStringValue(groupTrait.Profile, "application_id")
+	if !ok {
+		return nil, fmt.Errorf("databricks-connector: failed to get application_id from service principal profile")
+	}
+
+	ruleSets, err := s.client.ListRuleSets(ctx, ServicePrincipalsType, applicationId)
+	if err != nil {
+		return nil, fmt.Errorf("databricks-connector: failed to list rule sets for service principal %s (%s): %w", principal.Id.Resource, applicationId, err)
+	}
+
+	if len(ruleSets) == 0 {
+		l.Info(
+			"databricks-connector: service principal already does not have the entitlement",
+			zap.String("principal_id", principal.Id.Resource),
+			zap.String("entitlement", entitlement.Slug),
+		)
+
+		return nil, nil
+	}
+
+	principalID, err := preparePrincipalID(ctx, s.client, principal.Id.ResourceType, principal.Id.Resource)
+	if err != nil {
+		return nil, fmt.Errorf("databricks-connector: failed to prepare principal id: %w", err)
+	}
+
+	for i, ruleSet := range ruleSets {
+		if ruleSet.Role == entitlement.Slug {
```

Comment:
Another place where you can invert the logic to reduce indentation for the rest of the code in the loop.

```suggestion
		if ruleSet.Role != entitlement.Slug {
			continue
		}
```